python/src/worker/tasks/factory.py

```diff
@@ -36,39 +36,57 @@ class TaskFactory:
             GetDoubletScore,
             GetMitochondrialContent,
             MarkerHeatmap,
-            GetExpressionCellSets
+            GetExpressionCellSets,
         )
     }
 
     def __init__(self):
         self.count_matrix = CountMatrix()
         self.count_matrix.sync()
 
+    def _log_exception(self, task, error_object):
+        trace = (
+            f"Exception for task {task.__class__.__name__}:\n"
+            f"{traceback.format_exc()}"
+        )
+
+        error(trace)
+        xray_recorder.current_segment().add_exception(error_object, trace)
+
+        return trace
+
     def submit(self, msg):
         task = self._factory(msg)
 
         # Try to perform task. If fails, send back an error to the API.
         try:
             result = task.compute()
             return result
-        except Exception as e:
-            trace = f"Exception for task {task.__class__.__name__}:\n" \
-                    f"{traceback.format_exc()}"
 
-            error(trace)
-            xray_recorder.current_segment().add_exception(e, trace)
+        except Exception as e:
+            trace = self._log_exception(task, e)
 
-            # Only send real traces in development.
             if config.CLUSTER_ENV == "development":
                 result = Result(trace, error=True)
-            else:
-                result = [
-                    Result(
-                        "An unexpected error occurred while performing the work.",
-                        error=True,
-                    )
-                ]
-            return result
+
+            if isinstance(e, WorkerException):
```

Comment:
instead of the `isinstance` better to rely on the `except` mechanism (as you already simplified the handling by creating the `self._log_exception(task, e)`. e.g.:
```
except WorkerException as e:
            _ = self._log_exception(task, e)
            return Result(
                    {
                        "error_code": e.error_code,
                        "user_message": e.user_message,
                    },
                    error=True,
                )

except Exception as e:
            trace = self._log_exception(task, e)

            if config.CLUSTER_ENV == "development": # I think we can just remove this as in development we will see the error in the worker very easily
                return Result(trace, error=True)

            return Result(
                {
                    "error_code": "PYTHON_WORKER_GENERIC_ERROR",
                    "user_message": "An unexpected error occurred while performing the work.",
                },
                error=True,
            )
 
```

Reply:
Moved